### PR TITLE
Remove MLActivation parameters used for op fusion

### LINKED ISSUE #658: Consider removing `MLActivation` parameters used for op fusion
`MLActivation` currently has two uses:

1. With operators like `batchNormalization` and `conv2d`, as a hint that this activation may be fusable with the given operator
2. With recurrent operators like `lstm` and `gru`, as activations applied to gates within the given recurrent unit

| Operator           | Recurrent? | Purpose         |
| ------------------ | ---------- | --------------- |
| batchNormalization | No         | Maybe fusable   |
| conv2d             | No         | Maybe fusable   |
| convTranspose2d    | No         | Maybe fusable   |
| gru                | Yes        | Applied to gate |
| gruCell            | Yes        | Applied to gate |
| lstm               | Yes        | Applied to gate |
| lstmCell           | Yes        | Applied to gate |

I have some thoughts on (2), but I'll leave that to another issue. Let's focus on (1) here :)

My claim is that **there's no benefit to passing `MLActivation`s as parameters to `MLOperand`s for the purpose of op fusion**. Here's why:

1. *False positives*: For any given permutation of operator, activation, and backend, op fusion may not be possible (it's quite unlikely, actually)
   - CoreML does not natively support fusing activations with any of these operators
   - DirectML only supports fusing some activations with some operators. There's [an existing Chromium bug](https://crbug.com/336589268) to _un-fuse_ `MLActivation`s from their respective `MLOperand` if the combo is not supported for the given version of DML
2. *False negatives*: For any given operator which does _not_ take currently take an `MLActivation` in the WebNN spec, it may in fact be fusible with its input or output
   - [DML claims to support op fusion](https://learn.microsoft.com/en-us/windows/ai/directml/dml-fused-activations#operators-that-support-fused-activation) for `DML_OPERATOR_ELEMENT_WISE_ADD1` and `DML_OPERATOR_GEMM`, which ~map to WebNN's `add` and `gemm` operators
   - [TFLite supports fusing various operators](https://www.tensorflow.org/lite/models/convert/operation_fusion), as well

What this means in practice is:

- Implementations wrapping backends which can't fuse operators, or can't fuse some operators with some activations, must trivially break apart the `MLOperand` and its `MLActivation` into what's effectively just two `MLOperand`s connected to each other ([as Chromium's CoreML backend currently does](https://crrev.com/c/5459105)):
  ```
     input
       |
    operator
       |
   activation
       |
     output
  ```
- Implementations wrapping backends which _can_ fuse operators must do an optimization pass anyways to fuse operators which do not have `MLActivation` parameters in WebNN ([as Chromium's DML backend currently does](https://crsrc.org/c/services/webnn/dml/graph_impl.cc;l=4976-4990;drc=c00dbc9dc0167c5cf373db46f987869654c393ab))

Whether a given operator can be fused with a given activation is a very backend-specific quirk. Presumably we don't want to plumb through a new `MLActivation` operator to the web for every operator which any backend decides it can now fuse with some activation! This seems best left as an implementation detail either by the user agent (as described above) or the framework (who knows how much op fusion is happening in Core ML under the hood?! 🤔)

Thoughts?

### PR DESCRIPTION
**Pull request by a-sully**
Fixes #658

***
<a href="https://pr-preview.s3.amazonaws.com/a-sully/webnn/pull/664.html" title="Last updated on May 3, 2024, 11:58 PM UTC (082ceb3)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/664/4165154...a-sully:082ceb3.html" title="Last updated on May 3, 2024, 11:58 PM UTC (082ceb3)">Diff</a>

### CODE REVIEW

**Review by inexorabletash — CHANGES_REQUESTED** [review]

File: index.bs [hunk 1]
```diff
@@ -1557,19 +1549,18 @@ partial interface MLGraphBuilder {
...
-    return builder.relu(
-      builder.add(
-        builder.mul(
-          builder.reshape(options.scale, shape),
-          builder.div(
-            builder.sub(input, builder.reshape(mean, shape)),
-            builder.sqrt(builder.add(builder.reshape(variance, shape), builder.constant(options.epsilon)))
-            )),
-        builder.reshape(options.bias, shape)));
+    builder.add(
```
[the remark this answers is not in the document]
Comment (a-sully) [reply index.bs]: Whoops, didn't mean to remove the `return`. Added back

**Review by inexorabletash — APPROVED** [review]

**Review by fdwr — APPROVED** [review]
Looks good to me Mr. Sully. Thanks for opening the issue and updating the spec.

We should get a collateral impact assessment from @Honry and @huningxin, as I see [them used](https://github.com/microsoft/onnxruntime/blob/f9febc4f359371f9c3dd523f8e80bbe757471e33/onnxruntime/core/providers/webnn/builders/impl/normalization_op_builder.cc#L82) in [a few places](https://github.com/microsoft/onnxruntime/blob/f9febc4f359371f9c3dd523f8e80bbe757471e33/onnxruntime/core/providers/webnn/builders/impl/conv_op_builder.cc#L141), but it would be mostly *code removal and simplification* in the WebNN EP, as the [Chromium implementation](https://chromium-review.googlesource.com/c/chromium/src/+/5222281) fuses them anyway regardless of the caller.

File: index.bs [hunk 2]
```diff
@@ -2047,10 +2029,6 @@ partial interface MLGraphBuilder {
     : <dfn>bias</dfn>
     ::
         An additional 1-D tensor with the shape of *[outputChannels]* whose values are to be added to the convolution result.
-
-    : <dfn>activation</dfn>
-    ::
-        An optional activation function that immediately follows the convolution operation.
```
[the remark this answers is not in the document]
Comment (a-sully) [reply index.bs:2053]: To be honest I don't think we should be adding notes about op fusion for any specific operators, since which operators can be fused is backend-specific. I've instead expanded on the high-level discussion of op fusion. Thoughts?

File: index.bs [hunk 3]
```diff
@@ -1186,14 +1186,14 @@ interface MLActivation {};
 </div>
 
 <div class="note">
-These activations function types are used to create other operations. One such use of this interface is for when an activation function is fused into another operation such as {{MLGraphBuilder/conv2d()}} or {{MLGraphBuilder/batchNormalization()}} during a graph construction session. Such fused activation functions can provide a significant performance improvement when supported natively by the underlying implementation. This is intended as an optimization opportunity for implementers.
```
[the remark this answers is not in the document]
Comment (a-sully) [reply index.bs:1189]: Added a high-level note about graph construction above. I don't think this discussion should be in the `MLActivation` section :)

File: index.bs [hunk 4]
```diff
@@ -651,7 +651,7 @@ that shares the same buffer as the input tensor. (In the case of reshape,
 the entire data is shared, while in the case of slice, a part of the input data is shared.)
 The implementation may use views, as above, for intermediate values.
 
-Before the execution, the computation graph that is used to compute one or more specified outputs needs to be compiled and optimized. The key purpose of the compilation step is to enable optimizations that span two or more operations, such as operation or loop fusion.
+Before the execution, the computation graph that is used to compute one or more specified outputs needs to be constructed, compiled, and optimized. The key purpose of the compilation step is to enable optimizations that span two or more operations, such as operation or loop fusion. The user agent may also perform these optimizations during graph construction.
```
Comment (huningxin) [inline index.bs:654]: Does it make sense to use term "conversion"? That would refer to the procedure which converts the WebNN graph to native graph, rather than the "construction" of WebNN graph by JS code. This hopefully would align with the "converting" step of [`build()`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-build) method

> Let graphImpl be the result of converting [this](https://webidl.spec.whatwg.org/#this)'s [graph](https://www.w3.org/TR/webnn/#mlgraphbuilder-graph) with operands, operators, inputs, and outputs’s [values](https://infra.spec.whatwg.org/#map-getting-the-values) into an [implementation-defined](https://infra.spec.whatwg.org/#implementation-defined) format which can be interpreted by the underlying platform.

I suppose the user agent could perform these optimizations at that step, like Chromium DirectML does.

```suggestion
Before the execution, the computation graph that is used to compute one or more specified outputs needs to be converted, compiled, and optimized. The key purpose of the compilation step is to enable optimizations that span two or more operations, such as operation or loop fusion. The user agent may also perform these optimizations during graph conversion.
```

File: index.bs (hunk 2 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:2053]: I agreed we should put it on the high-level discussion. op fusion is platform specific, and we won't want to cover every single op.

File: index.bs (hunk 3 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:1189]: SGTM

**Comment by a-sully** [comment]
Thanks for the suggestions! They've since been incorporated :)

**Review by fdwr — APPROVED** [review]

**Review by huningxin — APPROVED** [review]
LGTM!
